op-program/client/interop/consolidate.go

```diff
@@ -33,36 +37,87 @@ func ReceiptsToExecutingMessages(depset depset.ChainIndexFromID, receipts ethtyp
 	return execMsgs, logCount, nil
 }
 
-func RunConsolidation(deps ConsolidateCheckDeps,
-	oracle l2.Oracle,
+func fetchAgreedBlockHashes(oracle l2.Oracle, superRoot *eth.SuperV1) ([]common.Hash, error) {
+	agreedBlockHashes := make([]common.Hash, len(superRoot.Chains))
+	for i, chain := range superRoot.Chains {
+		output := oracle.OutputByRoot(common.Hash(chain.Output), chain.ChainID)
+		outputV0, ok := output.(*eth.OutputV0)
+		if !ok {
+			return nil, fmt.Errorf("unsupported L2 output version: %d", output.Version())
+		}
+		agreedBlockHashes[i] = common.Hash(outputV0.BlockHash)
+	}
+	return agreedBlockHashes, nil
+}
+
+func RunConsolidation(
+	logger log.Logger,
+	bootInfo *boot.BootInfoInterop,
+	l1PreimageOracle l1.Oracle,
+	l2PreimageOracle l2.Oracle,
 	transitionState *types.TransitionState,
 	superRoot *eth.SuperV1,
+	tasks taskExecutor,
 ) (eth.Bytes32, error) {
+	deps, err := newConsolidateCheckDeps(transitionState, superRoot.Chains, l2PreimageOracle)
+	if err != nil {
+		return eth.Bytes32{}, fmt.Errorf("failed to create consolidate check deps: %w", err)
+	}
+
 	var consolidatedChains []eth.ChainIDAndOutput
 
+	agreedBlockHashes, err := fetchAgreedBlockHashes(l2PreimageOracle, superRoot)
+	if err != nil {
+		return eth.Bytes32{}, err
+	}
+
 	for i, chain := range superRoot.Chains {
 		progress := transitionState.PendingProgress[i]
 
-		// TODO(#13776): hint block data execution in case the pending progress is not canonical so we can fetch the correct receipts
-		block, receipts := oracle.ReceiptsByBlockHash(progress.BlockHash, chain.ChainID)
+		// It's possible that the optimistic block is not canonical.
+		// So we use the blockDataByHash hint to trigger a block rebuild to ensure that the block data, including receipts, are available.
+		_ = l2PreimageOracle.BlockDataByHash(agreedBlockHashes[i], progress.BlockHash, chain.ChainID)
+
+		optimisticBlock, receipts := l2PreimageOracle.ReceiptsByBlockHash(progress.BlockHash, chain.ChainID)
 		execMsgs, _, err := ReceiptsToExecutingMessages(deps.DependencySet(), receipts)
 		if err != nil {
 			return eth.Bytes32{}, err
 		}
 
 		candidate := supervisortypes.BlockSeal{
 			Hash:      progress.BlockHash,
-			Number:    block.NumberU64(),
-			Timestamp: block.Time(),
+			Number:    optimisticBlock.NumberU64(),
+			Timestamp: optimisticBlock.Time(),
 		}
+		consolidatedOutputRoot := progress.OutputRoot
 		if err := checkHazards(deps, candidate, chain.ChainID, execMsgs); err != nil {
-			// TODO(#13776): replace with deposit-only block if ErrConflict, ErrCycle, or ErrFuture
-			return eth.Bytes32{}, err
+			if !isInvalidMessageError(err) {
+				return eth.Bytes32{}, err
+			}
+			chainAgreedPrestate := superRoot.Chains[i]
+			optimisticBlockOutput := l2PreimageOracle.OutputByRoot(common.Hash(progress.OutputRoot), chain.ChainID)
```

Comment:
@ajsutton I just realized from working on reorg action tests, that the host doesn't have the required optimistic output root preimage. The host only knows the prestate super root preimage, but not the output root preimage of each chain.
The supervisor has this info, so we'll need the challenger to route that to the host configuration.

Reply:
Basically, can't we do: https://github.com/ethereum-optimism/optimism/blob/7719c5a582198de5827dc6b2501808e891ecc1cd/op-program/client/l2/engine.go#L41-L57

but with GetHeaderByHash instead of Number?

Reply:
Thanks for the suggestion. Could you please take a look at the fix [here](https://github.com/ethereum-optimism/optimism/pull/13826/commits/66c0fe00393674204a217aa260bc10019ff4efbf).